### PR TITLE
fix: typescript module init in existing project

### PR DESCRIPTION
**Pull request by TomChv**
This commit improve the entrypoint registration:
- Use `reflect-metadata` to register modules' class constructor.

This commit includes several changes around the runtime:
- Add a proxy entrypoint to execute dagger entrypoint as a sub-package.
- Install the TypeScript SDK in the project `package.json` if required.
- Update project `tsconfig.json` if the paths alias is missing.
- Add `src/index.ts` if missing in the codebase.
- Simplify template install `dagger` as a sub package.

All theses changes makes possible the initialization of a module inside an existing project.

This commit includes new tests on `module_test.go` on these features.

Fixes DEV-3301
Related to DEV-3147
Related to DEV-3029

### CODE REVIEW

File: sdk/typescript/.changes/unreleased/Added-20240117-125553.yaml
Comment (jedevc) [inline sdk/typescript/.changes/unreleased/Added-20240117-125553.yaml:1]: Since modules are still experimental, we don't need to release note this.

File: sdk/typescript/runtime/template/package.json
Comment (jedevc) [inline sdk/typescript/runtime/template/package.json:9]: Why is this a `devDependencies`? From my memory (not a js dev, so sorry):
- `dependencies` is for runtime requirements
- `devDependencies` is for compilation requirements

**Review by jedevc — COMMENTED** [review]
Will defer to @helderco on the actual changes to the typescript SDK here, since I'm not familiar with this part of the code, but from the tests it looks like the behavior is correct!

File: sdk/typescript/.changes/unreleased/Added-20240117-125553.yaml
Comment (TomChv) [reply sdk/typescript/.changes/unreleased/Added-20240117-125553.yaml:1]: Oh that's write, I forget again that we don't publish release note for Zenith related changes

Comment (TomChv) [reply sdk/typescript/.changes/unreleased/Added-20240117-125553.yaml:1]: Done in [`e1b9f18` (#6433)](https://github.com/dagger/dagger/pull/6433/commits/e1b9f1808949df952a922f7e6cd2b1a9d8763f11)

File: sdk/typescript/runtime/template/package.json
Comment (helderco) [reply sdk/typescript/runtime/template/package.json:9]: It's more appropriate to be in `dependencies` when considering that _production_ is the runtime container. But for the IDE, you only want `devDependencies`.

Since the runtime container is calling `npm install`, which installs from `dependencies` and `devDependencies`, it works in both cases.

However, locally, ideally, you install via a link so that changes to the module code are immediate. Otherwise, it may require you to call `npm install` on the host whenever you make changes that affect the codegen.

In Python, the runtime container installs manually from `pip install /sdk`, but locally, you should use an editable install with `pip install -e ./sdk` for those reasons.

Comment (TomChv) [reply sdk/typescript/runtime/template/package.json:9]: > However, locally, ideally, you install via a link so that changes to the module code are immediate. Otherwise, it may require you to call npm install on the host whenever you make changes that affect the codegen.

You cannot change `sdk` locally since it's dynamically injected by the runtime so any local change in that folder would be undo (except if you update dagger engine like with `./hack/dev`)

Comment (TomChv) [reply sdk/typescript/runtime/template/package.json:9]: > It's more appropriate to be in dependencies when considering that production is the runtime container.

It's true and false at the same time, you might have your CI setup in your root project with you other dependencies.
In that case, you don't want to compile your web app or nestJS app with dagger installed in it, since dagger shall only be executed through `dagger` cli so it's already wrapped into the right environment etc

**Comment by helderco** [comment]
Just a couple questions, waiting for reply but otherwise should be good to go.

Concerned about the requirement for `npm install` after `dagger mod use` but can be addressed in a follow up if needed.

File: sdk/typescript/runtime/template/package.json
Comment (helderco) [reply sdk/typescript/runtime/template/package.json:9]: > You cannot change `sdk` locally since it's dynamically injected by the runtime so any local change in that folder would be undo (except if you update dagger engine like with `./hack/dev`)

That's not what I meant. If you add a dependency (module), you'll need to rerun codegen. If you don't link, you'll have to `npm install` again locally after `dagger mod use`. That's why I said "whenever you make changes that affect the codegen". It's something that changes `./sdk`.

File: sdk/typescript/entrypoint/register.ts
Comment (TomChv) [reply sdk/typescript/entrypoint/register.ts:8]: It doesn't compile if I omit this :(

File: sdk/typescript/runtime/main.go
Comment (TomChv) [reply sdk/typescript/runtime/main.go:114]: I should have remove this `yarn.lock` actually, we shouldn't choose one for the user

File: sdk/typescript/runtime/main.go
Comment (TomChv) [reply sdk/typescript/runtime/main.go:135]: Ok I'll remove that

File: sdk/typescript/tsconfig.json
Comment (TomChv) [reply sdk/typescript/tsconfig.json:29]: You're right

File: sdk/typescript/runtime/template/package.json
Comment (TomChv) [reply sdk/typescript/runtime/template/package.json:9]: > That's not what I meant. If you add a dependency (module), you'll need to rerun codegen. If you don't link, you'll have to npm install again locally after dagger mod use. That's why I said "whenever you make changes that affect the codegen". It's something that changes ./sdk.

Actually not, the `tsconfig.json` is the one in charge of the inferance, you shouldn't need to npm install.
I'll record an exemple to verify my assumption

Comment (TomChv) [reply sdk/typescript/runtime/template/package.json:9]: https://github.com/dagger/dagger/assets/61683879/777e8d0c-59aa-4df3-a1d8-c63f17178000

@helderco See that I'm installing a module, I do not `npm install` but still the auto completion works and type safety as well (even found I forgot the `Promise`) :D

File: sdk/typescript/entrypoint/register.ts
Comment (helderco) [reply sdk/typescript/entrypoint/register.ts:8]: Sorry, I'm not suggesting a change here 🙂 Just that it reminded me of Node's creator regretting having support for **not putting** the extension (youtube link above), like it was before. That was for `require()` though, because it makes node go look into the various files that could match this (is it .js, .ts, etc...?).

File: sdk/typescript/runtime/main.go
Comment (helderco) [reply sdk/typescript/runtime/main.go]: With a node_modules cache volume that's another story 🙂

File: sdk/typescript/runtime/template/package.json
Comment (helderco) [reply sdk/typescript/runtime/template/package.json:9]: Good, thanks for the confirmation, it's good to know.

**Review by jedevc — APPROVED** [review]
:tada: Will be great to get this in v0.9.7!

**Pull request merged**
